lib/cc/analyzer/include_paths_builder.rb

```diff
@@ -0,0 +1,156 @@
+module CC
+  module Analyzer
+    class IncludePathsBuilder
+      def initialize(cc_exclude_paths)
+        @cc_exclude_paths = cc_exclude_paths
+      end
+
+      def build
+        root = Directory.new('.', ignored_files)
+        paths = root.included_paths
+        paths.each do |path|
+          raise_on_unreadable_files(path)
+        end
+        paths
+      end
+
+      protected
+
+      def ignored_files
+        Tempfile.open(".cc_gitignore") do |tmp|
+          tmp.write(File.read(".gitignore")) if File.exist?(".gitignore")
+          tmp << @cc_exclude_paths.join("\n")
+          tmp.close
+          tracked_and_ignored = `git ls-files -zi -X #{tmp.path}`.split("\0")
+          untracked_and_ignored = `git ls-files -zio -X #{tmp.path}`.split("\0")
+          tracked_and_ignored + untracked_and_ignored
+        end
+      end
+
+      def included_files
+        includable_files - ignored_files
+      end
+
+      def includable_files
+        tracked_in_git = `git ls-files -z`.split("\0")
+        untracked_in_git = `git ls-files -zo`.split("\0")
+        tracked_in_git + untracked_in_git
+      end
+
+      def raise_on_unreadable_files(path)
+        if File.directory?(path)
+          raise_on_unreadable_files_in_directory(path)
+        elsif !FileUtils.readable_by_all?(path)
+          raise(CC::Analyzer::UnreadableFileError, "Can't read #{path}")
+        end
+      end
+
+      def raise_on_unreadable_files_in_directory(path)
+        raw_entries = Dir.entries(path).reject do |e|
+          %w(. .. .git).include?(e)
+        end
+        raw_entries.each do |entry|
+          sub_path = File.join(path, entry)
+          raise_on_unreadable_files(sub_path)
+        end
+      end
+
+      class Directory
+        def initialize(path, excluded_files)
+          @path = path
+          @excluded_files = ensure_hashified(excluded_files)
+        end
+
+        def all_included?
+          readable_by_all? &&
+            files_all_included? &&
+            subdirectories_all_included?
+        end
+
+        def included_paths
+          if all_included?
+            [@path + "/"]
+          elsif readable_by_all?
+            result = []
+            result = result.concat(included_file_entries)
+            result = result.concat(included_subdirectory_results)
+            result
+          else
+            []
+          end
+        end
+
+        protected
+
+        def ensure_hashified(obj)
+          if obj.is_a?(Array)
+            result = {}
+            obj.each do |included|
+              result[included] = true
+            end
+            result
```

Comment:
You could replace the above 5 lines with something like this:

```ruby
obj.each_with_object({}) do |included, result|
  result[included]
end
```

and it'll return the resulting hash at the end.